### ISSUE TITLE
Softmax behaviour when used on sparse matrix

### ISSUE THREAD

**Issue by Fed03** [rusty1s/pytorch_geometric#2911]
Hi,
Probably a silly question but I've a doubt on how `torch_geometric.utils.softmax`/`torch_sparse.scatter_softmax` works when applied to a sparse matrix.

Ex given a vector of values and an indices tensor
```python
val = torch.as_tensor([3,7,9,8,1]).float()
indices = torch.as_tensor([
    [1,0,2,0,1],
    [0,2,1,1,1]
])
```
that is a 3x3 matrix
```python
matr = torch.as_tensor([
    [0,8,7],
    [3,1,0],
    [0,9,0]
]).float()
```
I want to calc softmax row_wise. If I apply torch.softmax to the dense matrix, I obtain
```python
torch.softmax(matr, dim=1)

tensor([[4.5279e-02, 2.6888e-01, 9.9818e-01],
        [9.0944e-01, 2.4518e-04, 9.1022e-04],
        [4.5279e-02, 7.3088e-01, 9.1022e-04]])
```
I don't understand how to use the `torch_geometric.utils.softmax`/`torch_sparse.scatter_softmax` with the former definition

**Comment by rusty1s** [comment]
```python
softmax(val, indices[0])
```
will compute the `softmax` row-wise, i.e., based on the indices in `indices[0]`. Note that `zero` values in the dense 3x3 matrix will be interpreted as `-inf` though.

**Comment by Fed03** [comment]
Many thanks.

This means that, for example, the last row would be `[0,1,0]` instead of `[4.5279e-02, 7.3088e-01, 9.1022e-04]`.
I don't understand if this can be considered correct. If the goula is to pass the softmax to an argmax function then it makes no difference, but in other application like self-attention I don't think we can use this interpretation.

What do you think?

**Comment by rusty1s** [comment]
There are two different options for computing the softmax in a sparse scenario:
* Interpret missing values as 0
* Interpret missing values as "undefined"
We take the second approach, as in attention-based GNNs, we only want to compute the softmax across neighborhoods (independent of the number of nodes).

**Comment by Fed03** [comment]
Understood, thanks! I apologize for my silly question ☺️

**Comment by rusty1s** [comment]
That wasn't a silly question at all. It's good aware of these differences in computation in comparison to a dense computation.
